rustls/examples/internal/test_ca.rs

```diff
@@ -0,0 +1,365 @@
+                .unwrap();
+            let CertifiedKey {
+                cert: end_entity, ..
+            } = certified_keys
+                .get(&(role, alg))
+                .unwrap();
+            fn write_chain_file(
+                role: Role,
+                alg: &'static SignatureAlgorithm,
+                name: &str,
+                chain_certs: &[&Certificate],
+            ) {
+                let mut chain_file =
+                    File::create(output_directory(alg).join(format!("{}.{}", role.label(), name)))
+                        .unwrap();
+                for cert in chain_certs {
+                    chain_file
+                        .write_all(cert.pem().as_bytes())
+                        .unwrap();
+                }
+            }
+
+            write_chain_file(role, alg, "chain", &[intermediate, root]);
+            write_chain_file(role, alg, "fullchain", &[end_entity, intermediate, root]);
+        }
+    }
+
+    // Write PEM CRLs revoking the client, end entity and intermediate certs.
+    for role in [Role::Intermediate, Role::EndEntity, Role::Client] {
+        for alg in sig_algs {
+            let CertifiedKey { cert, .. } = certified_keys
+                .get(&(role, alg))
+                .unwrap();
+
+            // The CRL will be signed by the issuer of the certificate being revoked. For
+            // intermediates this will be the trust anchor, and for client/EE certs this will
+            // be the intermediate.
+            let (issuer, issuer_key) = match role {
+                Role::Intermediate => {
+                    let CertifiedKey {
+                        cert: issuer,
+                        key_pair: issuer_key,
+                    } = certified_keys
+                        .get(&(Role::TrustAnchor, alg))
+                        .unwrap();
+                    (issuer, issuer_key)
+                }
+                Role::EndEntity | Role::Client => {
+                    let CertifiedKey {
+                        cert: issuer,
+                        key_pair: issuer_key,
+                    } = certified_keys
+                        .get(&(Role::Intermediate, alg))
+                        .unwrap();
+                    (issuer, issuer_key)
+                }
+                _ => panic!("unexpected role for CRL generation: {role:?}"),
+            };
+            let crl = crl_for_serial(
+                cert.params()
+                    .serial_number
+                    .as_ref()
+                    .unwrap(),
+            )
+            .signed_by(issuer, issuer_key)
+            .unwrap();
+            let mut crl_file = File::create(
+                output_directory(alg).join(format!("{}.revoked.crl.pem", role.label())),
+            )
+            .unwrap();
+            crl_file
+                .write_all(crl.pem().unwrap().as_bytes())
+                .unwrap();
+        }
+    }
+}
+
+struct CertifiedKey {
```

Comment:
Why not use `rcgen::CertifiedKey` for this?

Reply:
Done.